Avoid conflicts between default [semi-]voiced sound mark rules and custom rules

kzrnm opened the pull request:
Fixes #55008, Fixes #55009

**Background**

Let's take `は` as an example. Other characters have the same mechanism.

Default collation rules has `&は<<<ハ<<ば<<<バ<<ぱ<<<パ`. 

If strength is less than tertiary, the default rule can be regarded as `&は=ハ<<ば=バ<<ぱ=パ`.
In the current implementation, a custom rule like `&は<ハ&ば<バ&ぱ<パ` is added.
Merging these rules, it will be **`&は<<ば<<ぱ<パ<バ<ハ`**.

**This PR implimentation**

I change the custom rule to `&は<ハ`.
Merged rule is **`&は<<ば<<ぱ<ハ<<バ<<パ`**.


This PR don't fix small hiragana/katakana order issue which is #54987.

**reference**

- http://www.unicode.org/reports/tr35/tr35-collation.html#Orderings

kzrnm committed:
Remove custom dakuten rules

Avoid conflicts between default dakuten rules
and custom rules.

tarekgh commented on src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c:
```diff
@@ -158,11 +166,28 @@ static void FillIgnoreKanaRules(UChar* completeRules, int32_t* fillIndex, int32_
+        {
+            // Hiragana is the range 3041 to 3096 & 309D & 309E
+            if (hiraganaChar <= 0x3096 ||
+                hiraganaChar >= 0x309D) // characters between 3096 and 309D are not mapped to katakana
```
nit: please return these check to be in one line. thanks!

tarekgh commented:
>     UChar compareChar = isIgnoreKanaType ? '=' : '<';

now we already split the code accordign to isIgnoreKanaType so we can just get rid of this line and have compareChar initalized inside the if (isIgnoreKanaType) statement below. 

---
Refers to: src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c:161 in 58f62d0. [](commit_id = 58f62d072fbba0ec80f5e85df71fcc07bf1e0d2b, deletion_comment = False)

tarekgh commented on src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c:
```diff
@@ -112,6 +112,14 @@ static const UChar g_HalfFullHigherChars[] = {
 };
 static const int32_t g_HalfFullCharsLength = (sizeof(g_HalfFullHigherChars) / sizeof(UChar));
 
+static const UChar g_HiraganaWithoutDakutenChars[] = {
```
>g_HiraganaWithoutDakutenChars [](http://example.com/codeflow?start=19&length=29)

nit: please add some comment here telling why we specicial case this range and how it will be used.

tarekgh commented:
@naminodarie thanks for providing the fix. This is great!

I added a minore comments, other than that looks good.

kzrnm committed:
Apply review feedback

ufcpp commented on src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c:
```diff
@@ -112,11 +112,14 @@ static const UChar g_HalfFullHigherChars[] = {
 };
 static const int32_t g_HalfFullCharsLength = (sizeof(g_HalfFullHigherChars) / sizeof(UChar));
 
+// Hiragana without dakuten or handakuten for custom collation rules
```
For non-Japanese speaker, it might be better to use voiced sound/semi-voiced sound instead of dakuten/handakuten?

naminodarie replied:
The character names in unicode is not dakuten/handakuten but voiced sound/semi-voiced sound mark. 
That's a good point. 

|code point|name|
|---|---|
|3099|COMBINING KATAKANA-HIRAGANA VOICED SOUND MARK|
|309A|COMBINING KATAKANA-HIRAGANA SEMIVOICED SOUND MARK|
|309B|KATAKANA-HIRAGANA VOICED SOUND MARK|
|309C|KATAKANA-HIRAGANA SEMI-VOICED SOUND MARK|

naminodarie replied:
I have rewritten dakuten/handakuten to [semi-]voiced sound mark.

kzrnm committed:
Rewrite to [semi-]voiced sound mark

The name in unicode is not dakuten/handakuten but voiced
sound/semi-voiced sound mark.
